### PR TITLE
chore: update all emptyDirs to 'medium: Memory'

### PR DESCRIPTION
**Pull request by gabemontero**
@jhutar @pmacik @amfred @psuriset 

I've mentioned this in some of the WG calls.  We should compare the numbers from https://github.com/redhat-appstudio/build-definitions/pull/514 with this change

I'm curious how alleviating the use of aws storage for the sbom-json-check during the adv pipeline in particular has on overall disk IO

fwiw I don't see the use of the s2i-nodejs task in our runs

but ideally we pull the trigger on all our emptydirs at this point

### CODE REVIEW

File: task/buildah/0.1/buildah.yaml
Comment (lcarva) [inline task/buildah/0.1/buildah.yaml:335]: Any concerns with this approach when performing a large build?

Comment (gabemontero) [reply task/buildah/0.1/buildah.yaml:335]: good question @lcarva .... we dove into this a bit during the Aug 29 arch call.  The answer is "no" because
- the build team already has had to establish choices wrt the memory limits for the build container (stemming from hive's onboarding); per what you see in https://github.com/redhat-appstudio/build-definitions/tree/main/task we already have 4gb, 6gb, 8gb, 10gb choices, where more can be added
- the sandbox / RHTAP use of the cluster autoscaler should allocate more nodes on demand based on requests for more memory; 
- memory has been our least constrained resource to date
- memory is cheaper than storage wrt AWS
- the percentage delta per build from our stress runs was reasonable, especially when compared to the performance/scaling improvment

**Comment by gabemontero** [comment]
Finally captured the e2e failure before it got pruned:

```
[build-service-suite Build templates E2E test] HACBS pipelines triggers PipelineRun for component with source URL https://github.com/redhat-appstudio-qe/devfile-sample-python-basic [build, HACBS, pipeline, build-templates-e2e]
/github.com/redhat-appstudio/e2e-tests/tests/build/build_templates.go:119
  "level"=0 "msg"="Waited for 1.038529439s due to client-side throttling, not priority and fairness, request: GET:https://172.30.0.1:443/apis/generators.external-secrets.io/v1alpha1?timeout=32s\n"
  repository test-app-551-Dyf2k-finish-own status request to github: 200
• [55.434 seconds]
------------------------------
[build-service-suite Build templates E2E test] HACBS pipelines should eventually finish successfully for component with Git source URL https://github.com/redhat-appstudio-qe/devfile-sample-python-basic [build, HACBS, pipeline, build-templates-e2e]
/github.com/redhat-appstudio/e2e-tests/tests/build/build_templates.go:135
  [PANICKED] in [It] - /usr/local/go/src/runtime/panic.go:260 @ 09/01/23 14:12:37.208
  timing - /github.com/redhat-appstudio/e2e-tests/pkg/framework/failure_report.go:26 @ 09/01/23 14:12:37.208
    Test started at 2023-09-01 14:12:37.200998604 +0000 UTC m=+55.495416995
    Test ended at 2023-09-01 14:12:37.208606313 +0000 UTC m=+55.503024685
• [PANICKED] [0.008 seconds]
[build-service-suite Build templates E2E test] HACBS pipelines [It] should eventually finish successfully for component with Git source URL https://github.com/redhat-appstudio-qe/devfile-sample-python-basic [build, HACBS, pipeline, build-templates-e2e]
/github.com/redhat-appstudio/e2e-tests/tests/build/build_templates.go:135

  [PANICKED] Test Panicked
  In [It] at: /usr/local/go/src/runtime/panic.go:260 @ 09/01/23 14:12:37.208

  runtime error: invalid memory address or nil pointer dereference

  Full Stack Trace
    github.com/redhat-appstudio/e2e-tests/tests/build.glob..func2.1.4()
    	/github.com/redhat-appstudio/e2e-tests/tests/build/build_templates.go:138 +0xbd
------------------------------
```

Could be related to https://github.com/redhat-appstudio/e2e-tests/pull/688 which merged 8 hours ago

@jsmid1 @psturc is it possible that given the code block 
```
			testNamespace = os.Getenv(constants.E2E_APPLICATIONS_NAMESPACE_ENV)
			if len(testNamespace) > 0 {
				asAdminClient, err := kubeapi.NewAdminKubernetesClient()
				Expect(err).ShouldNot(HaveOccurred())
				kubeadminClient, err = framework.InitControllerHub(asAdminClient)
				Expect(err).ShouldNot(HaveOccurred())
				_, err = kubeadminClient.CommonController.CreateTestNamespace(testNamespace)
				Expect(err).ShouldNot(HaveOccurred())
			} else {
				f, err = framework.NewFramework(utils.GetGeneratedNamespace("build-e2e"))
				Expect(err).NotTo(HaveOccurred())
				testNamespace = f.UserNamespace
				Expect(f.UserNamespace).NotTo(BeNil())
				kubeadminClient = f.AsKubeAdmin
			}
```

that was not changed by your PR, when we run these tests in build-definitions, that we go down the path such that `framework.NewFramework` is not called, and the `AsKubeAdmin` from `f.AsKubeAdmin.TektonController` is nil, and thus the panic and line 138?  The other ref `kubeadminClient.HasController` was called 2 lines above, so that is not a candidate for the nil reference.

**Comment by gabemontero** [comment]
opened https://issues.redhat.com/browse/RHTAPBUGS-746 @jsmid1 @psturc for ^^

**Comment by lcarva** [comment]
@gabemontero, I noticed this on my PR as well: https://github.com/redhat-appstudio/e2e-tests/pull/743. It includes a fix.

**Comment by gabemontero** [comment]
> @gabemontero, I noticed this on my PR as well: [redhat-appstudio/e2e-tests#743](https://github.com/redhat-appstudio/e2e-tests/pull/743). It includes a fix.

cool thanks @lcarva :-) !!

**Comment by psturc** [comment]
the build of [pip component](https://github.com/cachito-testing/pip-e2e-test) failed on
```
unshare: sigprocmask unblock failed: Invalid argument
```

<details>
  <summary>Click for full log</summary>

```bash
STEP-BUILD

Trying to pull registry.access.redhat.com/ubi9/python-39@sha256:01fe0c6b483a3c425a1b851569b4f6445244c95e16c195a4f52c6e860e82891b...
Getting image source signatures
Checking if image destination supports signatures
Copying blob sha256:fb76de9d1083d62946d5e43034624205b358b77720539b7ee130cdeb60500101
Copying blob sha256:fc07e4fdbabe54e68d395dead6e5f6c44318493d34abba496630ed6bfd19931f
Copying blob sha256:94a4c2c1356a9e01d6c107308546c37e6dcb7330b4682b70a18750bffddb5e0d
Copying blob sha256:25ead466f550b0bc693a7ce9debd7c18fa65a7d6986be207da566a72f0923f10
Copying config sha256:42ec795691e3fa0dc53e1d5d8a29a6db51bab96be3428011d5e95746759eb0fc
Writing manifest to image destination
Storing signatures
42ec795691e3fa0dc53e1d5d8a29a6db51bab96be3428011d5e95746759eb0fc
Build will be executed with network isolation
Prefetched content will be made available
STEP 1/10: FROM registry.access.redhat.com/ubi9/python-39@sha256:01fe0c6b483a3c425a1b851569b4f6445244c95e16c195a4f52c6e860e82891b
STEP 2/10: RUN . /cachi2/cachi2.env &&     if curl -IsS www.google.com; then echo "Has network access!"; exit 1; fi
curl: (6) Could not resolve host: www.google.com
time="2023-09-04T11:06:32Z" level=warning msg="pkg/chroot: error unmounting \"/var/tmp/buildah2561281872/mnt/rootfs\" (retried 0 times): permission denied"
time="2023-09-04T11:06:32Z" level=warning msg="pkg/bind: error detaching \"/var/tmp/buildah2561281872/mnt/buildah-bind-target-11\": permission denied"
time="2023-09-04T11:06:32Z" level=warning msg="pkg/bind: error removing \"/var/tmp/buildah2561281872/mnt/buildah-bind-target-11\": device or resource busy"
time="2023-09-04T11:06:32Z" level=warning msg="pkg/bind: error detaching \"/var/tmp/buildah2561281872/mnt/rootfs\": permission denied"
time="2023-09-04T11:06:32Z" level=warning msg="pkg/bind: error removing \"/var/tmp/buildah2561281872/mnt/rootfs\": device or resource busy"
time="2023-09-04T11:06:32Z" level=warning msg="pkg/bind: error removing \"/var/tmp/buildah2561281872/mnt\": directory not empty"
STEP 3/10: WORKDIR /opt/test_package_cachi2
STEP 4/10: COPY . .
STEP 5/10: RUN . /cachi2/cachi2.env &&     pip install -r requirements.txt
Looking in links: /cachi2/output/deps/pip
Processing /cachi2/output/deps/pip/external-dockerfile-parse/dockerfile-parse-external-sha256-36e4469abb0d96b0e3cd656284d5016e8a674cd57b8ebe5af64786fe63b8184d.tar.gz
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Processing /cachi2/output/deps/pip/certifi-2022.12.7.tar.gz
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Processing /cachi2/output/deps/pip/charset-normalizer-3.0.1.tar.gz
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Processing /cachi2/output/deps/pip/idna-3.4.tar.gz
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Processing /cachi2/output/deps/pip/requests-2.28.2.tar.gz
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Processing /cachi2/output/deps/pip/urllib3-1.26.14.tar.gz
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Using legacy 'setup.py install' for certifi, since package 'wheel' is not installed.
Using legacy 'setup.py install' for charset-normalizer, since package 'wheel' is not installed.
Using legacy 'setup.py install' for urllib3, since package 'wheel' is not installed.
Using legacy 'setup.py install' for dockerfile-parse, since package 'wheel' is not installed.
Building wheels for collected packages: idna, requests
  Building wheel for idna (pyproject.toml): started
  Building wheel for idna (pyproject.toml): finished with status 'done'
  Created wheel for idna: filename=idna-3.4-py3-none-any.whl size=61539 sha256=6414764f630a0f1f4d776d8fad94ed8d9f0baef5320c8b20ecd938ae96751e19
  Stored in directory: /tmp/pip-ephem-wheel-cache-w7ru4er7/wheels/47/63/00/cf4278b3805c252509cddab7c8757262b6a8084a5e91282cca
  Building wheel for requests (pyproject.toml): started
  Building wheel for requests (pyproject.toml): finished with status 'done'
  Created wheel for requests: filename=requests-2.28.2-py3-none-any.whl size=62821 sha256=b3576c9f7e5c6fdd261a5922d1ae81bc7014c75f1700b5ac37e937d75fdec789
  Stored in directory: /tmp/pip-ephem-wheel-cache-w7ru4er7/wheels/f9/a7/28/94cd4e171a4b97baa549684beaf629e13f642c7c281860690e
Successfully built idna requests
Installing collected packages: urllib3, idna, charset-normalizer, certifi, requests, dockerfile-parse
    Running setup.py install for urllib3: started
    Running setup.py install for urllib3: finished with status 'done'
    Running setup.py install for charset-normalizer: started
    Running setup.py install for charset-normalizer: finished with status 'done'
    Running setup.py install for certifi: started
    Running setup.py install for certifi: finished with status 'done'
    Running setup.py install for dockerfile-parse: started
    Running setup.py install for dockerfile-parse: finished with status 'done'
Successfully installed certifi-2022.12.7 charset-normalizer-3.0.1 dockerfile-parse-2.0.0 idna-3.4 requests-2.28.2 urllib3-1.26.14
time="2023-09-04T11:06:46Z" level=warning msg="pkg/chroot: error unmounting \"/var/tmp/buildah3868955576/mnt/rootfs\" (retried 0 times): permission denied"
time="2023-09-04T11:06:46Z" level=warning msg="pkg/bind: error detaching \"/var/tmp/buildah3868955576/mnt/buildah-bind-target-11\": permission denied"
time="2023-09-04T11:06:46Z" level=warning msg="pkg/bind: error removing \"/var/tmp/buildah3868955576/mnt/buildah-bind-target-11\": device or resource busy"
time="2023-09-04T11:06:46Z" level=warning msg="pkg/bind: error detaching \"/var/tmp/buildah3868955576/mnt/rootfs\": permission denied"
time="2023-09-04T11:06:46Z" level=warning msg="pkg/bind: error removing \"/var/tmp/buildah3868955576/mnt/rootfs\": device or resource busy"
time="2023-09-04T11:06:46Z" level=warning msg="pkg/bind: error removing \"/var/tmp/buildah3868955576/mnt\": directory not empty"
STEP 6/10: CMD ["python", "/opt/test_package_cachi2/src/test_package_cachi2/main.py"]
STEP 7/10: LABEL "build-date"="2023-09-04T11:06:30"
STEP 8/10: LABEL "architecture"="x86_64"
STEP 9/10: LABEL "vcs-type"="git"
STEP 10/10: LABEL "vcs-ref"="1ecda839ba9ca55070d75c86c26a1bb07d777bba"
COMMIT quay.io/redhat-user-workloads/build-templates-e2e/test-app-551/pip-e2e-test-zfps:build-45263-1693825522
unshare: sigprocmask unblock failed: Invalid argument
```
</details>

**Comment by jhutar** [comment]
Hello @psturc . Pls how is this related to change in this PR? Any idea?

**Comment by gabemontero** [comment]
the retest failure was pruned before I could look @jhutar @psturc but from the man pages `unshare` definitely can have filesystem related applications.  It could be that making one of these volumes read only has had some unexpected side effect.  Especially if main branch in consistently clean now.

Now, "why?" and "how could we avoid?" are questions I don't know the answer to immediately.  I'm totally unfamiliar with how these RHTAP python / devfile builds.  Someone would have to dig into it.

We should also compare with https://github.com/redhat-appstudio/build-definitions/pull/514 and see if the same problem occurs there.  If not, and if we can at least get the buildah container change in, that is at least a start.

**Comment by gabemontero** [comment]
ok passes with 8G ... will now work back and see how low it can go

**Comment by gabemontero** [comment]
OK with 6G ... trying 5G

**Comment by gabemontero** [comment]
And it works with 5g mem limit

**Comment by gabemontero** [comment]
> FYI my attempts to update the 6b,8gb,10gb buildah tasks to patch `inject-sbom-and-push` failed in the e2e in ways I don't understand
> 
> we'll need to circle back to this after @jhutar verifies the default buildah task in a staging run

@jhutar saw a massive improvement with bumping `inject-sbom-and-push`  ... only a handful of failures, and they were 500s from `registry.access.redhat.com` ... we hit it too fast / hard ;-) 

so we need to circle back to getting the buildah 6gb,86b,10gb tasks bumped accordingly for their inject push steps, and this should become a "mergeable candidate" for consideration.

**Comment by gabemontero** [comment]
the e2e's passed this time

I suspect I may have hit some capacity issues with the `build-templates-e2e` namespace on prod-rh when I ran the tests yesterday.....I'm the only one running PRs in the build-definitions repo as I type

I'll do some retests earlier in the day tomorrow, get a larger test sample size, but I suspect 
@jhutar @psuriset @amfred @mmorhun @arewm @adambkaplan we are ready to sync up here and make the final decision to merge this change

ICYMI - this switch to memory for `/var/lib/containers` is providing significant performance/scaling improvements in the performance WG test runs in staging.

**Comment by adambkaplan** [comment]
I recommend that the commits are squashed, and update the commit message to provide the justifications for the change. We'll want a record in git that explains the performance improvement and the impact of using "Memory" as the medium for emptyDir (increased memory resource requirements).

Net effect here is that we're using a "ramdisk" - @amfred is there a downstream impact here of provisioning more RAM-heavy nodes on the member clusters?

**Comment by gabemontero** [comment]
> I recommend that the commits are squashed, and update the commit message to provide the justifications for the change. We'll want a record in git that explains the performance improvement and the impact of using "Memory" as the medium for emptyDir (increased memory resource requirements).
> 

thanks @adambkaplan - I am waiting on a one line summary from @jhutar on the final perf/scaling benefits, and then I'll squash / reword the commits accordingly

and I'm still curious about the limits in the `build-templates-e2e` namespace per my last comment; I'll be retesting this PR a bit during the day wrt that

> Net effect here is that we're using a "ramdisk" - @amfred is there a downstream impact here of provisioning more RAM-heavy nodes on the member clusters?

correct re: ramdisk

your question is certainly pertinent; fortunately, it came up in the perf wg and has been vetted; that said, I'll let @amfred speak to it, so we have her official "buy in" documented here

thanks

**Comment by gabemontero** [comment]
quick side bar - turns our there are no `ResourceQuotas` or `ClusterResourceQuotas` defined for the `build-templates-e2e` namespace where e2e's for this repo run in (where those e2e's get launched from `tekton-ci` namespace)

**Review by amfred — APPROVED** [review]
Our performance/load tests are succeeding with this change, and it makes the builds run significantly faster, so let's do it.

File: task/buildah/0.1/buildah.yaml
Comment (mmorhun) [inline task/buildah/0.1/buildah.yaml:305]: Do we need it in production?

**Review by mmorhun — APPROVED** [review]
Technically LGTM.

**Comment by amfred** [comment]
> Net effect here is that we're using a "ramdisk" - @amfred is there a downstream impact here of provisioning more RAM-heavy nodes on the member clusters?

We did change the type of worker node that we use on our autoscaling pools to either m6i.2xlarge or m6id.2xlarge.  We did some tests to compare different node types to see which are faster overall.

Before this change we were only using around 25% of the memory on our nodes before they got full for other reasons, so actually using more memory is a good thing in this case.  It might make the autoscaler kick in sooner than it would have otherwise, but we're OK with that.  In practice we're only seeing 1-3 extra nodes during heavy load, which isn't too many.

**Comment by gabemontero** [comment]
I'm about to update the commits with this data, but I'm going to post here as well.  The latest memory findings from @jhutar are in particular interesting (this was new news we just got today, even before @amfred posted here comment ^^)

1) 42% reduction in workload duration and 80% improvement in amount of workloads finished within business defined time constrainst using builds of a NodeJS repos.

2) memory usage interestingly ends up being a wash:
- across the cluster, "Shmem" memory (memory used for RAM disks) went from 0 to 28 GB
- however, across the cluster, "Cached" memory (used by the disk's write-through cache and shared memory) dropped by 29 GB

File: task/buildah/0.1/buildah.yaml
Comment (gabemontero) [reply task/buildah/0.1/buildah.yaml:305]: yep need to fix that

**Comment by gabemontero** [comment]
commits squashed and commit message updated @adambkaplan 

use of debug logging in the `buildah push` removed @mmorhun 

PTAL and let's iterate toward tagging this for merging

thanks

**Comment by adambkaplan** [comment]
LGTM!

Once CI tests pass we definitely want to merge this. Great write up in the commit message.

**Comment by gabemontero** [comment]
CI is green @adambkaplan since the commit msg update but something else merged in the interim and it is asking for a rebase (man I miss how we auto-rebased PRs in OCP ;-) ) 

rebasing now .... I technically have merge authority here in this repo, and can merge this if need be, but ideally you or someone else from the build team hits the button after it comes up green again

thanks

**Comment by gabemontero** [comment]
FYI everyone we had to revert this PR with https://github.com/redhat-appstudio/build-definitions/pull/612

Once actual users started using this, we ran into way too many memory constraints, even with our bumping of the default.
Bottom line, all our testing still was not sufficient.

We'll need to either make this an opt in feature @arewm @adambkaplan @jhutar @psuriset @amfred or go back to figuring out how we can better estimate what the default size should be.

**Comment by sbose78** [comment]
> we ran into way too many memory constraints

Could you point me to the issues, please?

**Comment by gabemontero** [comment]
> > we ran into way too many memory constraints
> 
> Could you point me to the issues, please?

I'm working on a RCA doc now @sbose78 .... I'll make sure you have a link to it.  It will include my initial thoughts on how we might pursue this in the future.

But bottom line, simple configuration changes like we tried to do here are not generically viable IMO.

**Pull request merged**
